File: src/main/java/de/sean/splugin/spigot/tasks/SleepChecker.java
package de.sean.splugin.spigot.tasks;

/* SPlugin */
import de.sean.splugin.SPlugin;
import de.sean.splugin.util.SMessages;
import de.sean.splugin.util.SUtil;

/* Spigot */
import org.bukkit.Bukkit;
import org.bukkit.GameMode;
import org.bukkit.World;
import org.bukkit.entity.Entity;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;

/* Java */
import java.util.ArrayList;
import java.util.List;

public class SleepChecker implements Runnable {
    public static final List<World> skippingWorlds = new ArrayList<>();

    @Override
    public void run() {
        Bukkit.getOnlinePlayers().stream().map(Entity::getWorld).distinct().filter(this::validateWorld).forEach(this::checkWorld);
    }

    private boolean validateWorld(final World world) {
        // Time between 12950 and 23950 is night time.
        return !skippingWorlds.contains(world) && (world.getTime() > 12950L || world.getTime() < 23950L);
    }

    private void checkWorld(final World world) {
        final List<Player> players = world.getPlayers();
        final int amountIgnored = (int) players.stream().filter(SleepChecker::ignorePlayer).count();
        final int amountSleeping = (int) players.stream().filter(LivingEntity::isSleeping).count();
        final int amountNeeded = Math.max(0, (int)Math.ceil((players.size() - amountIgnored) * (SPlugin.instance.getConfig().getInt("skipNight.percentage") / 100.0) - amountSleeping));
        if (amountNeeded == 0 && amountSleeping > 0) {
            // More than 50% of players are sleeping
            players.forEach(player -> SMessages.sendActionBarMessage(player, SMessages.getRandomMessage("messages.everyoneSleeping")));
            skippingWorlds.add(world);
            new SkipNightTask(world).runTaskTimer(SPlugin.instance, 0L, 1L);
            players.forEach(player -> player.sendMessage(SMessages.getRandomMessage("messages.skipNight")));
        } else if (amountNeeded > 0 && amountSleeping > 0) {
            final String message = SMessages.getRandomMessage("messages.sleeping").replace("[sleeping]", Integer.toString(amountSleeping)).replace("[needed]", Integer.toString(amountNeeded));
            players.forEach(player -> SMessages.sendActionBarMessage(player, message));
        }
    }

    private static boolean ignorePlayer(final Player player) {
        // We will ignore AFK players and players who are not in survival
        return player.getGameMode() != GameMode.SURVIVAL || SUtil.isPlayerAFK(player.getUniqueId());
    }
}

File: src/main/kotlin/de/sean/splugin/spigot/tasks/SleepChecker.kt
package de.sean.splugin.spigot.tasks

import de.sean.splugin.SPlugin
import de.sean.splugin.util.SMessages.getRandomMessage
import de.sean.splugin.util.SMessages.sendActionBarMessage
import de.sean.splugin.util.SUtil
import org.bukkit.Bukkit
import org.bukkit.GameMode
import org.bukkit.World
import org.bukkit.entity.Entity
import org.bukkit.entity.Player
import java.util.*
import java.util.function.Consumer
import kotlin.math.ceil

class SleepChecker : Runnable {
    override fun run() {
        Bukkit.getOnlinePlayers().stream().map { obj: Entity -> obj.world }.distinct().filter { world: World -> validateWorld(world) }.forEach { world: World -> checkWorld(world) }
    }

    private fun validateWorld(world: World): Boolean {
        // Time between 12950 and 23950 is night time.
        return !skippingWorlds.contains(world) && (world.time > 12950L || world.time < 23950L)
    }

    private fun checkWorld(world: World) {
        val players = world.players
        val amountIgnored = players.stream().filter { player: Player -> ignorePlayer(player) }.count().toInt()
        val amountSleeping = players.stream().filter { obj: Player -> obj.isSleeping }.count().toInt()
        val amountNeeded = Math.max(0, ceil(Math.max(0, players.size - amountIgnored) * (SPlugin.instance.config.getInt("skipNight.percentage") / 100.0) - amountSleeping).toInt())
        if (amountNeeded == 0 && amountSleeping > 0) {
            // More than 50% of players are sleeping
            players.forEach(Consumer { player: Player? -> sendActionBarMessage(player!!, getRandomMessage("messages.everyoneSleeping")) })
            skippingWorlds.add(world)
            SkipNightTask(world).runTaskTimer(SPlugin.instance, 0L, 1L)
            players.forEach(Consumer { player: Player -> player.sendMessage(getRandomMessage("messages.skipNight")) })
        } else if (amountNeeded > 0 && amountSleeping > 0) {
            val message = getRandomMessage("messages.sleeping").replace("[sleeping]", amountSleeping.toString()).replace("[needed]", amountNeeded.toString())
            players.forEach(Consumer { player: Player? -> sendActionBarMessage(player!!, message) })
        }
    }

    companion object {
        val skippingWorlds: MutableList<World> = ArrayList()
        private fun ignorePlayer(player: Player): Boolean {
            // We will ignore AFK players and players who are not in survival
            return player.gameMode != GameMode.SURVIVAL || SUtil.afkPlayers[player.uniqueId]!!
        }
    }
}
